## 5. Métodos iterativos

Os métodos iterativos são indicados para sistemas de grande porte com a matriz dos coeficientes esparsa, ou seja, com grande grande quantidade de zeros. Nesses casos os métodos iterativos são mais rápidos e demandam menos memória para armazenamento. Além disso, os métodos iterativos possuem a vantagem de se auto corrigir e também podem ser usados para reduzir os erros de arredondamento na solução obtida por métodos exatos. Sob certas condições, podem também ser usados para resolver um conjunto de equações não lineares. 

No entanto, para que o método iterativo seja convergente, é necessário atender a certas condições. Dizemos que a sequência de vetores $x^{(k)}$, $k=0,1,2,..$ de um espaço vetorial $V$ converge para a solução exata $\bar{x} \in V$ se $\|x^{(k)} - \bar{x}\|  \rightarrow 0$, quando $k \rightarrow \infty$ para alguma norma $ \| .\|$ sobre o espaço vetorial $V$.

Para determinar a solução de um sistema linear por métodos iterativos, precisamos transformar o sistema dado em um outro sistema equivalente que estabeleça um processo iterativo. Suponha então, que o sistema $Ax = b$ seja reescrito na forma equivalente 

$$x = Hx + g$$

tal que a solução $\bar{x}$ de $x = Hx + g$ é  também, solução de $Ax = b$. A partir dessa nova forma podemos obter um processo iterativo que fornecerá a sequência de soluções aproximadas que buscamos. Assim, seja $x^{(0)}$ uma aproximação inicial para a solução $\bar{x}$, obtemos as aproximações sucessivas $x^{(k)}$ para a solução desejada usando o processo iterativo definido por:

$$x^{(k)} = Hx^{(k-1)} + g$$

**Convergência:**
A condição necessária e suficiente para a convergência do processo iterativo definido por $x = Hx + g$ é que $max \{ |\lambda_i |\} < 1$, onde $\lambda_i$ são os autovalores da matriz $H$. 

**Critério geral de convergência:**
Como consequência do teorema acima, o processo será convergente se para qualquer norma de matrizes, $\| H \| < 1$.   

*Prova:* (Franco, 2006, p.169) Na k-ésima iteração o vetor erro, dado por $e^{(k)}=\bar{x}-x^{(k)}$ será 

$$ e ^{(k)} = H^ke^{(0)}$$

Tomando normas consistentes, tem-se que 

$$ \| H^k e^{(0)} \| \leq \| H \|^k \|e^{(0)}\|$$

portanto 
$$ \| e^{(k)} \| \leq \| H \|^k \|e^{(0)}\|$$

como $\| H \|<1$ tem-se que $\| e^{(k)} \| \rightarrow 0$ quando $k \rightarrow \infty$.

Assim, se $\| H \|<1$ para alguma norma, temos a garantia de que o processo iterativo dado por $ x - x^{(k)} = H(x-x^{(k-1)})$ converge para a solução exata $\bar{x}$. A matriz $H$ é chamada **matriz de iteração**.

### Método iterativo de Jacobi-Richardson

Considere um sistema  de equações lineares $Ax = b$ em que $det(A) \neq 0$, com a diagonal principal $a_{ii} \neq 0$, $i=1,...,n$ como segue

$$ \begin{cases} 
	         a_{11}x_1 +a_{12}x_2 + \cdots + a_{1n}x_n = b_1\\ 
	         a_{21}x_1 +a_{22}x_2 + \cdots + a_{2n}x_n = b_2\\
	         \vdots   \\
	         a_{n1}x_1 +a_{n2}x_2 + \cdots + a_{nn}x_n = b_1\\
             \end{cases} $$

Dividindo cada linha do sistema dado pelo elemento da diagonal e isolando $x_1$ na $1^a$ equação, $x_2$ na $2^a$ equação até $x_n$ na n-ésima equação, temos o sistema escrito na forma equivalente (contanto que não ocorram divisões por zero):

$$ \begin{cases} 
	         x_1 = \frac{1}{a_{11}} \left(b_1  - a_{12}x_2 - a_{13}x_3 - \cdots - a_{1n}x_n \right)\\ 
	         x_2 = \frac{1}{a_{22}} \left(b_2  - a_{21}x_1 - a_{23}x_3 - \cdots - a_{2n}x_n \right)\\ 
	         \vdots   \\
	         x_n = \frac{1}{a_{nn}} \left(b_n  - a_{n1}x_1 - a_{n2}x_2 - \cdots - a_{n \, n-1}x_{n-1} \right)\\ 
\end{cases} $$

O método iterativo de Jacobi-Richardson é dado da seguinte forma:

$$ \begin{cases} 
             x_1^{(k+1)} = \frac{b_1}{a_{11}}  - \frac{a_{12}}{a_{11}}x_2^{(k)} - \frac{a_{13}}{a_{11}}x_3^{(k)} - \cdots - \frac{a_{1n}}{a_{11}}x_n^{(k)} \\ 
	         x_2^{(k+1)} = \frac{b_2}{a_{22}}  - \frac{a_{21}}{a_{22}}x_1^{(k)} - \frac{a_{23}}{a_{22}}x_3^{(k)} - \cdots - \frac{a_{1n}}{a_{11}}x_n^{(k)} \\ 
	         \vdots   \\
	         x_n^{(k+1)} = \frac{b_n}{a_{nn}}  - \frac{a_{n1}}{a_{nn}}x_1^{(k)} - \frac{a_{n2}}{a_{nn}}x_3^{(k)} - \cdots - \frac{a_{n \, n-1}}{a_{nn}}x_{n-1}^{(k)} \\
\end{cases} $$


ou, ainda:

$$ x_i^{(k+1)} = \frac{1}{a_{ii}} \left(b_i - \sum_{j=1}^{i-1} a_{ij}x_j^{(k)} -\sum_{j=i+1}^{n} a_{ij}x_j^{(k)} \right) \,\,\,\,\,\, i=1,...,n$$

Na forma matricial, o método de Jacobi-Richardson pode ser escrito como:

$$ x^{(k+1)} = Hx^{(k)} + g $$

onde 


$$H = 
\left[\begin{array}{ccccc} 
 0 & -\frac{a_{12}}{a_{11}} & -\frac{a_{13}}{a_{11}} &\cdots & -\frac{a_{1n}}{a_{11}} \\ 
 -\frac{a_{21}}{a_{22}} & 0 & -\frac{a_{23}}{a_{22}} &\cdots & -\frac{a_{2n}}{a_{22}} \\
  \vdots & \vdots & \vdots & \vdots & \vdots \\
 -\frac{a_{n1}}{a_{nn}} & -\frac{a_{n2}}{a_{nn}} & -\frac{a_{n3}}{a_{nn}}&\cdots & 0 \\
\end{array} \right]
\quad \text{e} \quad
g = 
\left[\begin{array}{c} 
\frac{b_1}{a_{11}} \\ 
\frac{b_2}{a_{22}} \\
\vdots \\
\frac{b_n}{a_{nn}}\\
\end{array} \right]$$


ou, ainda,


$$
\left[\begin{array}{c} 
x_1^{(k+1)} \\ 
x_2^{(k+1)} \\
\vdots \\
x_n^{(k+1)}\\
\end{array} \right]
= 
\left[\begin{array}{ccccc} 
 0 & -\frac{a_{12}}{a_{11}} & -\frac{a_{13}}{a_{11}} &\cdots & -\frac{a_{1n}}{a_{11}} \\ 
 -\frac{a_{21}}{a_{22}} & 0 & -\frac{a_{23}}{a_{22}} &\cdots & -\frac{a_{2n}}{a_{22}} \\
  \vdots & \vdots & \vdots & \vdots & \vdots \\
 -\frac{a_{n1}}{a_{nn}} & -\frac{a_{n2}}{a_{nn}} & -\frac{a_{n3}}{a_{nn}}&\cdots & 0 \\
\end{array} \right]
\left[\begin{array}{c} 
x_1^{(k)} \\ 
x_2^{(k)} \\
\vdots \\
x_n^{(k)}\\
\end{array} \right]
+
\left[\begin{array}{c} 
\frac{b_1}{a_{11}} \\ 
\frac{b_2}{a_{22}} \\
\vdots \\
\frac{b_n}{a_{nn}}\\
\end{array} \right]
$$

#### Convergência
Podemos verificar a convergência do método Jacobi-Richardson avaliando a matriz $A$. Se a matriz $A=[a_{ij}]_{i,j=1.,,,,n}$ do sistema $Ax=b$ for estritamente diagonalmente dominante, ou seja, se

$$ |a_{ii}| > \sum_{j=1,i\neq j} |a_{ij}| , i=1,...n$$

então, o método iterativo será convergente. 

#### Critério de parada 
Considerando que o processo iterativo está fornecendo uma sequência convergente, um critério de parada para o algoritmo é dado por

$$ \frac{\parallel x^{(k)} - x^{(k-1)}\parallel }{\parallel x^{(k)}\parallel} < \epsilon$$

para alguma norma vetorial $\parallel  . \parallel : V \rightarrow R$ e alguma tolerância $\epsilon$ pré estabelecida. 

Por conveniência, é comum utilizarmos a norma infinito:

$$\parallel x\parallel _ \infty = max \{ |x_0|, |x_1|, ..., |x_n| \}$$

**Exemplo 5.1:**
Usando o método interativo de Jacobi-Richardson, determine uma solução aproximada para o seguinte sistema de equações lineares, com aproximação inicial $x^{(0)}=(0,0,0)^T$ e precisão $\epsilon = 0.01$. 

$$
\left[\begin{array}{ccc} 
10 & 2 & 1\\
1 & 5 & 1\\
2 & 3 & 10\\
\end{array} \right]
\left[\begin{array}{c} 
x_1\\
x_2\\
x_3\\
\end{array} \right]
=
\left[\begin{array}{c} 
14\\
11\\
8\\
\end{array} \right]
$$

Reescrevendo o sistema, obtemos:

$$ 
\begin{cases} 
x_1^{(k+1)} = \frac{14}{10} - \frac{2}{10}x_2^{(k)} - \frac{1}{10}x_3^{(k)}\\ 
x_2^{(k+1)} = \frac{11}{5} - \frac{1}{5}x_1^{(k)} - \frac{1}{5}x_3^{(k)}\\ 
x_3^{(k+1)} = \frac{8}{10} - \frac{2}{10}x_1^{(k)} - \frac{3}{10}x_2^{(k)}\\ 
\end{cases} 
$$

A partir da primeira iteração, $k=0$, $x^{(0)}=(x_1^{(0)},x_2^{(0)},x_3^{(0)})^T=(0,0,0)^T$obtem-se:

$x_1^{(1)}=\frac{14}{10}-\frac{2}{10}x_2^{(0)}-\frac{1}{10}x_3^{(0)}=\frac{14}{10}-\frac{2}{10}(0)-\frac{1}{10}(0)=1,4$


$x_2^{(1)}= \frac{11}{5}-\frac{1}{5}x_1^{(0)}-\frac{1}{5}x_3^{(0)}= \frac{11}{5}-\frac{1}{5}(0)-\frac{1}{5}(0)=2,2$


$x_3^{(1)}=\frac{8}{10}-\frac{2}{10}x_1^{(0)}-\frac{3}{10}x_2^{(0)}=\frac{8}{10}-\frac{2}{10}(0)-\frac{3}{10}(0)=0,8$

Para $k=0$ e tomando $x^{(0)}=(0,0,0)^t$, temos $x^{(1)}=(1,4;2,2;0,8)^t$
Teste de parada: $\frac{\left\|x^{(1)}-x^{(0)}\right\|_{\infty}}{\left\|x^{(1)}\right\|_{\infty}}=1>\varepsilon$

Para $k=1$, temos:

$x_1^{(2)}=\frac{14}{10}-\frac{2}{10} x_2^{(1)}-\frac{1}{10} x_3^{(1)}=\frac{14}{10}-\frac{2}{10} (2,2)-\frac{1}{10} (0,8)=0,88$

$x_2^{(2)}=\frac{11}{5}-\frac{1}{5} x_1^{(1)}-\frac{1}{5} x_3^{(1)} = \frac{11}{5}-\frac{1}{5} (1,4)-\frac{1}{5} (0,8) =1,76$

$x_3^{(2)}=\frac{8}{10}-\frac{2}{10} x_1^{(1)}-\frac{3}{10} x_2^{(1)}=\frac{8}{10}-\frac{2}{10} (1,4)-\frac{3}{10} (2,2)=-0,14$


Portanto,

$$
x^{(2)}=(0,8800;1,7600;-0,1400)^t, \quad \frac{\left\|x^{(2)}-x^{(1)}\right\|_{\infty}}{\left\|x^{(2)}\right\|_{\infty}}=0,5341>\varepsilon
$$


Assim, sucessivamente, calculamos $x^{(3)}$, $x^{(4)}$, $x^{(5)}$, $\dots$, até que o critério de parada seja satisfeito em

$$x^{(6)}=(0,9967;1,9955;-0,0047)^t$$

em que

$$ \frac{\left\|x^{(6)}-x^{(5)}\right\|_{\infty}}{\left\|x^{(6)}\right\|_{\infty}}=0,0091<\varepsilon$$

Em python podemos fazer como é mostrado a seguir

In [1]:
import numpy as np

In [2]:
# Dados iniciais
x = np.array([0.,0.,0.])

A =  np.array([[10.,2.,1.],
               [1.,5.,1.],
               [2.,3.,10.]])
b = np.array([14.,11.,8.])

In [3]:
# função para fazer as iterações
def itera(x):
    x1 = (b[0] - A[0,1]*x[1] - A[0,2]*x[2])/A[0,0]
    x2 = (b[1] - A[1,0]*x[0] - A[1,2]*x[2])/A[1,1]
    x3 = (b[2] - A[2,0]*x[0] - A[2,1]*x[1])/A[2,2]
    x = np.array([x1, x2, x3])
    return x

In [4]:
x = itera(x)
x

array([1.4, 2.2, 0.8])

In [5]:
#várias iterações e o erro
x = np.array([0.,0.,0.])

x_ant = x
for i in range(2,8):
    x = itera(x)
    err = np.max(abs(x-x_ant))/np.max(abs(x))
    x_ant = x
    print (i, x, np.round(err,4))
    
print ("Solução:", np.round(x,4))

2 [1.4 2.2 0.8] 1.0
3 [ 0.88  1.76 -0.14] 0.5341
4 [1.062 2.052 0.096] 0.1423
5 [ 0.98    1.9684 -0.028 ] 0.063
6 [1.00912 2.0096  0.01348] 0.0206
7 [ 0.996732  1.99548  -0.004704] 0.0091
Solução: [ 0.9967  1.9955 -0.0047]


Uma outra forma de obtermos os mesmo resultados usando a expressão resumida, seria

In [6]:
A = np.array([[10., 2., 1.],[1.,5.,1.],[2.,3.,10.]])
b = np.array([14.,11.,8.])
x = np.array([0.,0.,0.])
n = len(A)

xk = np.array([0.,0.,0.])
for k in range(6):
    for i in range(n):
        xk[i] = (b[i]-np.sum(A[i,0:i]*x[:i])-np.sum(A[i,i+1:]*x[i+1:]))/A[i,i]
        # a linha acima é equivalente a:
        #xk[i] = (b[i]-np.dot(A[i,0:i],x[:i])-np.dot(A[i,i+1:],x[i+1:]))/A[i,i]
    print (xk)
    x = xk.copy()

[1.4 2.2 0.8]
[ 0.88  1.76 -0.14]
[1.062 2.052 0.096]
[ 0.98    1.9684 -0.028 ]
[1.00912 2.0096  0.01348]
[ 0.996732  1.99548  -0.004704]


In [7]:
print ("Solução:", np.round(x,4))

Solução: [ 0.9967  1.9955 -0.0047]


Resolvendo novamente, mas agora usando operações matriciais

In [8]:
#Calculando a matriz H
H = A.copy()
g = b.copy()

for i in range(len(A)):
    H[i] = -A[i]/A[i,i]
    g[i] = b[i]/A[i,i]
    
H = H + np.identity(len(A))
print ('H =', H)
print ('g =', g)

H = [[ 0.  -0.2 -0.1]
 [-0.2  0.  -0.2]
 [-0.2 -0.3  0. ]]
g = [1.4 2.2 0.8]


In [9]:
#ou, alternativamente
H = np.eye(len(A))- A/np.diag(A).reshape(3,1)
print ('H =', H)

H = [[ 0.  -0.2 -0.1]
 [-0.2  0.  -0.2]
 [-0.2 -0.3  0. ]]


In [10]:
g = b/np.diag(A)
print ('g =', g)

g = [1.4 2.2 0.8]


In [11]:
# verificando a convergẽncia com a norma inf 
norma_inf = np.sum(abs(H), axis=1).max()
print ('Norma:', norma_inf)

Norma: 0.5


In [12]:
x = np.array([0,0,0])

In [13]:
x = H@x+g
x

array([1.4, 2.2, 0.8])

In [14]:
# fazendo as iterações
err = 1
x = np.array([0,0,0])
x_ant = x

print ('Iterações:')
while err>0.01:
    x = np.dot(H,x)+g
    err = abs(max(x-x_ant)/max(x))
    x_ant = x   
    print(np.round(x, 4), ", Err=", np.round(err,5))

Iterações:
[1.4 2.2 0.8] , Err= 1.0
[ 0.88  1.76 -0.14] , Err= 0.25
[1.062 2.052 0.096] , Err= 0.1423
[ 0.98    1.9684 -0.028 ] , Err= 0.04166
[1.0091 2.0096 0.0135] , Err= 0.02064
[ 0.9967  1.9955 -0.0047] , Err= 0.00621


**Atividade 1:** Use o método de Jacobi-Richardson, se possível, para achar a solução do sistema linear dado no Exercício 1, na última seção. 

#### Método iterativo de Gaus-Seidel

Considere um sistema  de equações lineares $Ax = b$ em que $det(A) \neq 0$, com a diagonal principal $a_{ii} \neq 0$, $i=1,...,n$ como segue

$$ \begin{cases} 
	         a_{11}x_1 +a_{12}x_2 + \cdots + a_{1n}x_n = b_1\\ 
	         a_{21}x_1 +a_{22}x_2 + \cdots + a_{2n}x_n = b_2\\
	         \vdots   \\
	         a_{n1}x_1 +a_{n2}x_2 + \cdots + a_{nn}x_n = b_1\\
             \end{cases} $$

Dividindo cada linha do sistema dado pelo elemento da diagonal e isolando $x_1$ na $1^a$ equação, $x_2$ na $2^a$ equação até $x_n$ na n-ésima equação, temos o sistema escrito na forma equivalente:

$$ \begin{cases} 
	         x_1 = \frac{1}{a_{11}} \left(b_1  - a_{12}x_2 - a_{13}x_3 - \cdots - a_{1n}x_n \right)\\ 
	         x_2 = \frac{1}{a_{22}} \left(b_2  - a_{21}x_1 - a_{23}x_3 - \cdots - a_{2n}x_n \right)\\ 
	         \vdots   \\
	         x_n = \frac{1}{a_{nn}} \left(b_n  - a_{n1}x_1 - a_{n2}x_2 - \cdots - a_{n \, n-1}x_{n-1} \right)\\ 
\end{cases} $$

O método iterativo de Gauss-Seidel é dado da seguinte forma:

$$ \begin{cases} 
             x_1^{(k+1)} = \frac{b_1}{a_{11}}  - \frac{a_{12}}{a_{11}}x_2^{(k)} - \frac{a_{13}}{a_{11}}x_3^{(k)} - \cdots - \frac{a_{1n}}{a_{11}}x_n^{(k)} \\ 
	         x_2^{(k+1)} = \frac{b_2}{a_{22}}  - \frac{a_{21}}{a_{22}}x_1^{(k+1)} - \frac{a_{23}}{a_{22}}x_3^{(k)} - \cdots - \frac{a_{1n}}{a_{11}}x_n^{(k)} \\ 
	         \vdots   \\
	         x_n^{(k+1)} = \frac{b_n}{a_{nn}}  - \frac{a_{n1}}{a_{nn}}x_1^{(k+1)} - \frac{a_{n2}}{a_{nn}}x_3^{(k+1)} - \cdots - \frac{a_{n \, n-1}}{a_{nn}}x_{n-1}^{(k+1)} \\
\end{cases} $$


De forma geral, o método de Gauss-Seidel pode ser resumido como:

$$ x_i^{(k+1)} = \frac{1}{a_{ii}} \left(b_i - \sum_{j=1}^{i-1} a_{ij}x_j^{(k+1)} -\sum_{j=i+1}^{n} a_{ij}x_j^{(k)} \right) \,\,\,\,\,\, i=1,...,n$$

ou ainda, de forma equivalente:

$$ x_i^{(k+1)} = g_i - \sum_{j=1}^{i-1} h_{ij}x_j^{(k+1)} -\sum_{j=i+1}^{n} h_{ij}x_j^{(k)} \,\,\,\,\,\, i=1,...,n$$

em que $g_i=\frac{b_i}{a_{ii}}$ e $h_{ij}=\frac{a_{ij}}{a_{ii}}$, $i=1,...,n$. 

#### Convergência: critério de Sassenfield
Sejam as constantes $\beta_i$ definidas pelas seguintes fórmulas de recorrência:

$$\beta_i = \sum_{j=1}^{i-1} |h_{ij}|\beta_i +\sum_{j=i+1}^{n} |h_{ij}| \,\,\,\,\,\, i=1,...,n$$

e seja

$$\beta = \max_{1 \leq i \leq n}\ \beta_i$$

Então, se $\beta < 1$, a sequência $x^{(k)}$, gerada pelo método iterativo de Gauss-Seidel, converge para a solução $x$ do sistema dado.

*Prova:* Ver em [Arenales e Darezzo (2016)](https://integrada.minhabiblioteca.com.br/reader/books/9788522112821/pageid/80). 

**Exemplo 5.2:** Usando o método iterativo de Gauss-Seidel, determine uma solução aproximada para o sistema dado a seguir, com aproximação inicial
$x^{(0)} =(x_1^{(0)},x_2^{(0)},x_3^{(0)})^t =(0,0,0)^t$ e precisão $\epsilon = 0.01$.

$$ \left[\begin{array}{ccc} 
10 & 2 & 1\\ 1 & 5 & 1\\ 2 & 3 & 10\\
\end{array} \right]
\left[\begin{array}{c} 
x_1\\ x_2\\ x_3\\
\end{array} \right]
=
\left[\begin{array}{c} 
14\\ 11\\ 8\\
\end{array} \right] $$

Construção da matriz H:

$$
\mathrm{H}=\left[\begin{array}{rrr}
0 & -2 / 10 & -1 / 10 \\
-1 / 5 & 0 & -1 / 5 \\
-2 / 10 & -3 / 10 & 0
\end{array}\right]
$$


Verificando a condição de convergência (critério de Sassenfeld):

$$
\begin{aligned}
& \beta_1=\sum_{j=2}^3\left|h_{i j}\right| \rightarrow \beta_1=3 / 10=0.3000 \\
& \beta_2=\left|h_{21}\right| \beta_1+\left|h_{23}\right| \rightarrow \beta_2=13 / 50=0.2600 \\
& \beta_3=\left|h_{31}\right| \beta_1+\left|h_{32}\right| \beta_2 \rightarrow \beta_3=69 / 500=0.1380
\end{aligned}
$$


Assim,

$$
\beta=\max _{1 \leq i \leq 3} \beta_i=\operatorname{máx}\{0.3000,0.2600,0.1380\}=0.3000<1
$$

Portanto, temos garantia da convergência da sequência de soluções aproximadas geradas pelo método iterativo de Gauss-Seidel.

Vejamos como fazer as contas usando Python.

In [15]:
# Entrando com a matriz a e o vetor b
A = np.array([[10.0, 2.0, 1.0],
              [ 1.0, 5.0, 1.0],
              [ 2.0, 3.0, 10.0]])

b = np.array([14., 11., 8.])

In [16]:
# obtendo a matriz H
H = np.eye(len(A))-(A.T/np.diag(A)).T

print ('Matriz H:')
print(H)

Matriz H:
[[ 0.  -0.2 -0.1]
 [-0.2  0.  -0.2]
 [-0.2 -0.3  0. ]]


In [17]:
# verificando a convergência 
beta = []
for i in range(len(H)):
    bi = np.dot(beta,abs(H[i,0:i]))+ np.sum(abs(H[i,i+1:]))
    beta.append(np.round(bi,4))

print('Convergência:')
print ('max{',beta,'} =',np.array(beta).max())

Convergência:
max{ [0.3, 0.26, 0.138] } = 0.3


In [18]:
n = len(A)
x = np.array([0.,0.,0.])

for k in range(1,5):
    for i in range(len(A)):
        x[i] = (b[i] - np.dot(A[i,0:i],x[0:i])-np.dot(A[i,i+1:n],x[i+1:n]))/A[i,i]
    print(k, np.round(x,4))

print("Solução:",np.round(x,4))

1 [ 1.4    1.92  -0.056]
2 [ 1.0216  2.0069 -0.0064]
3 [ 9.9930e-01  2.0014e+00 -3.0000e-04]
4 [0.9997 2.0001 0.    ]
Solução: [0.9997 2.0001 0.    ]


___
Na forma matricial, o método de Gauss-Seidel  pode ser escrito como

$$\left[
    \begin{array}{c} 
	         x_1^{(k+1)} \\ 
	         x_2^{(k+1)} \\
             \vdots\\
	         x_n^{(k+1)} \\
	\end{array} 
\right]
=
\left[\begin{array}{ccccc} 
	         0                     & 0 & 0 & \cdots & 0 \\ 
	         -\frac{a_{21}}{a_{22}} & 0 & 0 & \cdots & 0 \\
	         \vdots & \vdots & \vdots & \vdots & \vdots \\
	         -\frac{a_{n1}}{a_{nn}} & -\frac{a_{n2}}{a_{nn}} & \cdots & -\frac{a_{n \,n-1}}{a_{nn}} & 0\\
	         \end{array} \right]
             \left[
    \begin{array}{c} 
	         x_1^{(k+1)} \\ 
	         x_2^{(k+1)} \\
             \vdots\\
	         x_n^{(k+1)} \\
	\end{array} 
\right]
+
\left[\begin{array}{ccccc} 
	         0&  -\frac{a_{12}}{a_{11}}& -\frac{a_{13}}{a_{11}}& \cdots &-\frac{a_{1n}}{a_{11}} \\ 
	         0& 0& -\frac{a_{23}}{a_{22}} & \cdots & -\frac{a_{2n}}{a_{22}} \\
	         \vdots & \vdots & \vdots & \vdots & \vdots \\
	         0& 0& 0& \cdots& 0\\
	         \end{array} \right]
\left[\begin{array}{c} 
	         x_1^{(k)} \\ 
	         x_2^{(k)} \\
             \vdots\\
	         x_n^{(k)} \\
	\end{array} \right]
+
\left[ \begin{array}{c} 
	         \frac{b_1}{a_{11}} \\ 
	         \frac{b_2}{a_{22}} \\
             \vdots\\
	         \frac{b_n}{a_{nn}} \\
	\end{array} \right]$$


ou 
$$ x^{(k+1)} = P x^{(k+1)} + Q x^{(k)} + g$$

$$ (I-P)x^{(k+1)} = Q x^{(k)} + g$$

$$ x^{(k+1)} = (I-P)^{-1}Q x^{(k)} + (I-P)^{-1}g$$

Fazendo-se $H = (I-P)^{-1}Q$ e $g' =(I-P)^{-1}g$ o processo iterativo torna-se

$$x^{(k+1)} = H x^{(k)} + g'$$

**Exemplo 5.3:** Vamos resolver o Exemplo 5.2 usando Gauss-Seidel na forma matricial

In [19]:
from numpy.linalg import inv

x = np.array([0.,0.,0.])

eps = 0.001
# Entrando com a matriz a e o vetor b
A = np.array([[10.0, 2.0, 1.0],
              [ 1.0, 5.0, 1.0],
              [ 2.0, 3.0, 10.0]])

b = np.array([14., 11., 8.])

n = len(A)
P = np.zeros((n,n))
Q = np.zeros((n,n))
g = np.zeros(n)


for i in range(n):
    P[i,0:i] = -A[i,0:i]/A[i,i]
    Q[i,i+1:n] = -A[i,i+1:n]/A[i,i]
    g[i] = b[i]/A[i,i] 
    
I = np.eye(n)
H = np.dot(inv(I-P),Q)
g = np.dot(inv(I-P),g)


for k in range(4):
    x = np.dot(H,x) + g
    
    print (np.round(x,4))

[ 1.4    1.92  -0.056]
[ 1.0216  2.0069 -0.0064]
[ 9.9930e-01  2.0014e+00 -3.0000e-04]
[0.9997 2.0001 0.    ]


Ou, de um modo mais esperto...

In [20]:
P = np.tril(H)
P

array([[0.   , 0.   , 0.   ],
       [0.   , 0.04 , 0.   ],
       [0.   , 0.028, 0.074]])

In [21]:
Q = np.triu(H)
Q

array([[ 0.   , -0.2  , -0.1  ],
       [ 0.   ,  0.04 , -0.18 ],
       [ 0.   ,  0.   ,  0.074]])

In [22]:
Hl = np.linalg.inv(np.eye(len(H))-P)@Q
gl = np.linalg.inv(np.eye(len(H))-P)@g

In [23]:
# fazendo as iterações
err = 1
x = np.array([0,0,0])
x_ant = x

print ('Iterações:')
while err>0.001:
    x = (Hl@x)+gl
    err = abs(max(x-x_ant)/max(x))
    x_ant = x   
    print(np.round(x, 4), ", Err=", np.round(err,5))

Iterações:
[ 1.4  2.  -0. ] , Err= 1.0
[1.     2.0833 0.0025] , Err= 0.04
[0.9831 2.0863 0.0028] , Err= 0.00144
[0.9825 2.0864 0.0028] , Err= 3e-05


### Exercícios 

**1.** Usando algum método iterativo, determine uma solução aproximada para os sistemas dados, com aproximação inicial
$x^{(0)} =(0,0,...,0)^t$ e 10 iterações.

a)
$\left[\begin{array}{ccccc} 
5 & 1 & 2 & 1 & -1\\ 
0&6&1&1&-1\\ 
0&1&-3&2&1\\
3&0&2&7&0\\
1&1&0&0&8\\
\end{array} \right]
\left[\begin{array}{c} 
x_1\\ x_2\\ x_3\\x_4\\x_5\\
\end{array} \right]
=\left[\begin{array}{c} 
1\\ 2\\ 0\\2\\1\\
\end{array} \right] \quad \quad$
b)
$\begin{aligned} 
4 x_1+x_2+x_3+\quad x_5 & =6, \\ 
-x_1-3 x_2+x_3+x_4 & =6, \\ 
2 x_1+x_2+5 x_3-x_4-x_5 & =6, \\ 
-x_1-x_2-x_3+4 x_4 & =6, \\ 
2 x_2-x_3+x_4+4 x_5 & =6\end{aligned} \quad \quad$
c)
$\begin{array}{rlr}
4 x_1-x_2 & =0, \\ 
-x_1+4 x_2-x_3 & =5, \\ 
-x_2+4 x_3 & =0, \\ 
+4 x_4-x_5 & =6, \\ 
-x_4+4 x_5-x_6= & -2, \\ 
-x_5+4 x_6 & =6 .
\end{array}$

**2.** ([Burden et al, 2016, p.505](https://integrada.minhabiblioteca.com.br/reader/books/9788522123414/pageid/521)) Use a técnica iterativa de Gauss-Seidel para encontrar soluções aproximadas de

$$
\begin{aligned}
10 x_1-x_2+2 x_3 & =6, \\
-x_1+11 x_2-x_3+3 x_4 & =25, \\
2 x_1-x_2+10 x_3-x_4 & =-11, \\
3 x_2-x_3+8 x_4 & =15,
\end{aligned}
$$

começando com $\mathbf{x}=(0,0,0,0)^t \mathrm{e}$ iterando até que

$$
\frac{\left\|\mathbf{x}^{(k)}-\mathbf{x}^{(k-1)}\right\|_{\infty}}{\left\|\mathbf{x}^{(k)}\right\|_{\infty}}<10^{-3} .
$$

**3.** ([Chapra e Canale, 2016, p. 305](https://integrada.minhabiblioteca.com.br/reader/books/9788580555691/pageid/305))

Um problema comum na engenharia elétrica envolve a determinação das correntes e tensões em várias posições em circuitos de resistores.

Considere o curcuíto mostrado abaixo

<img src="https://raw.githubusercontent.com/tiagoburiol/MatComp_II/main/imagens/circuito.png" width="400">
A regra de corrente de Kirchhoff aplicada em cada nó do circuito abaixo fornece

$$
\begin{aligned}
& i_{12}+i_{52}+i_{32}=0 \\
& i_{65}-i_{52}-i_{54}=0 \\
& i_{43}-i_{32}=0 \\
& i_{54}-i_{43}=0
\end{aligned}
$$

A aplicação da regra de tensão em cada um dos dois laços dá

$$
\begin{aligned}
& -i_{54} R_{54}-i_{43} R_{43}-i_{32} R_{32}+i_{52} R_{52}=0 \\
& -i_{65} R_{65}-i_{52} R_{52}-i_{12} R_{12}-200=0
\end{aligned}
$$

Substituindo o valor das resistências e movendo as constantes para o lado direito, o problema se transforma em resolver o  conjunto de seis equações equações lineares para as seis correntes desconhecidas. 

Obtenha o sistema e encontre as correntes usando algum método de eliminação.


**4.** ([Chapra e Canale, 2016, p. 310](https://integrada.minhabiblioteca.com.br/reader/books/9788580555691/pageid/310))
Uma bomba peristáltica produz um escoamento unitário ($Q_i$) de um fluido altamente viscoso. A rede está descrita na figura abaixo. 

<img src="https://raw.githubusercontent.com/tiagoburiol/MatComp_II/main/imagens/bomba_peristaltica.png" width="300">

Cada seção de tubo tem os mesmos comprimento e diâmetro. Os balanços de massa e energia mecânica podem ser simplificados para se obter os escoamentos em cada tubo como solução do sistema de equações 

$$
\begin{aligned}
&\begin{aligned}
& Q_3+2 Q_4-2 Q_2=0 \\
& Q_5+2 Q_6-2 Q_4=0\\
& 3 Q_7-2 Q_6=0
\end{aligned}\\
&\begin{aligned}
& Q_1=Q_2+Q_3 \\
& Q_3=Q_4+Q_5 \\
& Q_5=Q_6+Q_7
\end{aligned}
\end{aligned}
$$

Resolva o sistema para obter o escoamento de cada corrente.

### Referências

ARENALES, Selma; DAREZZO, Artur. Cálculo Numérico: aprendizagem com Apoio de Software. 2. ed. Porto Alegre: +A Educação - Cengage Learning Brasil, 2016. 

BURDEN, Richard L.; FAIRES, J D.; BURDEN, Annette M. Análise Numérica - Tradução da 10ª edição norte-americana. 3. ed. Porto Alegre: +A Educação - Cengage Learning Brasil, 2016. 

CHAPRA, Steven C.; CANALE, Raymond P. Métodos numéricos para engenharia. 7. ed. Porto Alegre: AMGH, 2016.

FRANCO, Neide Bertoldi. Cálculo numérico. São Paulo: Pearson Prentice Hall, 2006. 

RUGGIERO, M. A. G. e LOPES, V. L. R. Cálculo Numérico: Aspectos Teóricos e Computacionais. 2ª edição. Saraiva, 2016.

